File: javapatterns/src/main/java/org/iut/nantes/factories/pizzas/abstractfactory/CreerIngredients.java
package org.iut.nantes.factories.pizzas.abstractfactory;

abstract class CreerIngredients {
    abstract Pate creerPate();
    abstract Sauce creerSauce();
    abstract Fromage creerFromage();
}

class IngredientsNantes extends CreerIngredients {
    @Override
    public Pate creerPate() {
        return new PateFine();
    }

    @Override
    public Sauce creerSauce() {
        return new SauceTomate();
    }

    @Override
    Fromage creerFromage() {
        return new Emmental();
    }
}

class IngredientsNY extends CreerIngredients {

    @Override
    public Pate creerPate() {
        return new PateEpaisse();
    }

    @Override
    public Sauce creerSauce() {
        return new CremeFraiche();
    }

    @Override
    Fromage creerFromage() {
        return new Cheddar();
    }
}

File: kotlinpatterns/src/main/kotlin/org/iut/nantes/factories/pizzas/abstractfactory/CreerIngredients.kt
package org.iut.nantes.factories.pizzas.abstractfactory

abstract class CreerIngredients {
    abstract fun creerPate(): Pate
    abstract fun creerSauce(): Sauce
    abstract fun creerFromage(): Fromage
}

class IngredientsNantes : CreerIngredients() {
    override fun creerPate(): Pate {
        return PateFine()
    }

    override fun creerSauce(): Sauce {
        return SauceTomate()
    }

    override fun creerFromage(): Fromage {
        return Emmental()
    }
}

class IngredientsNY : CreerIngredients() {
    override fun creerPate(): Pate {
        return PateEpaisse()
    }

    override fun creerSauce(): Sauce {
        return CremeFraiche()
    }

    override fun creerFromage(): Fromage {
        return Cheddar()
    }
}